phoenix-core-client/src/main/java/org/apache/phoenix/cache/ServerMetadataCacheImpl.java

```diff
@@ -0,0 +1,165 @@
+    private static final String PHOENIX_COPROC_REGIONSERVER_CACHE_TTL_MS =
+            "phoenix.coprocessor.regionserver.cache.ttl.ms";
+    // Keeping default cache expiry for 30 mins since we won't have stale entry
+    // for more than 30 mins.
+    private static final long DEFAULT_PHOENIX_COPROC_REGIONSERVER_CACHE_TTL_MS
+            = 30 * 60 * 1000L; // 30 mins
+    private static final String PHOENIX_COPROC_REGIONSERVER_CACHE_SIZE
+            = "phoenix.coprocessor.regionserver.cache.size";
+    private static final long DEFAULT_PHOENIX_COPROC_REGIONSERVER_CACHE_SIZE = 10000L;
+    private static volatile ServerMetadataCacheImpl cacheInstance;
+    private MetricsMetadataCachingSource metricsSource;
+
+    /**
+     * Creates/gets an instance of ServerMetadataCache.
+     *
+     * @param conf configuration
+     * @return cache
+     */
+    public static ServerMetadataCacheImpl getInstance(Configuration conf) {
+        ServerMetadataCacheImpl result = cacheInstance;
+        if (result == null) {
+            synchronized (ServerMetadataCacheImpl.class) {
+                result = cacheInstance;
+                if (result == null) {
+                    cacheInstance = result = new ServerMetadataCacheImpl(conf);
+                }
+            }
+        }
+        return result;
+    }
+
+    public ServerMetadataCacheImpl(Configuration conf) {
+        this.conf = HBaseConfiguration.create(conf);
+        this.metricsSource = MetricsPhoenixCoprocessorSourceFactory
+                                .getInstance().getMetadataCachingSource();
+        long maxTTL = conf.getLong(PHOENIX_COPROC_REGIONSERVER_CACHE_TTL_MS,
+                DEFAULT_PHOENIX_COPROC_REGIONSERVER_CACHE_TTL_MS);
+        long maxSize = conf.getLong(PHOENIX_COPROC_REGIONSERVER_CACHE_SIZE,
+                DEFAULT_PHOENIX_COPROC_REGIONSERVER_CACHE_SIZE);
+        lastDDLTimestampMap = CacheBuilder.newBuilder()
+                .removalListener((RemovalListener<ImmutableBytesPtr, Long>) notification -> {
+                    String key = notification.getKey().toString();
+                    LOGGER.debug("Expiring " + key + " because of "
+                            + notification.getCause().name());
+                })
+                // maximum number of entries this cache can handle.
+                .maximumSize(maxSize)
+                .expireAfterAccess(maxTTL, TimeUnit.MILLISECONDS)
+                .build();
+    }
+
+    /**
+     * Returns the last DDL timestamp from the table.
+     * If not found in cache, then query SYSCAT regionserver.
+     * @param tenantID tenant id
+     * @param schemaName schema name
+     * @param tableName table name
+     * @return last DDL timestamp
+     * @throws Exception
+     */
+    public long getLastDDLTimestampForTable(byte[] tenantID, byte[] schemaName, byte[] tableName)
+            throws SQLException {
+        String fullTableNameStr = SchemaUtil.getTableName(schemaName, tableName);
```

Comment:
This seems to be used only for logging. We could eliminate this call and the log itself or just log the byte[] formatted as strings.

Reply:
done